### PR TITLE
Tree equivalence

### PR DESCRIPTION
**Pull request by katiasuya**

### CODE REVIEW

**Comment by katiasuya** [comment]
I think I should handle the error if the user uses not int in New() function, should I?

File: tree_test1.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	ch := make(chan int)
+	go Walk(New(5, -9, 8, 0), ch)
+
+	for {
+		val, ok := <-ch
+		if !ok {
+			break
+		}
+		fmt.Print(" ", val)
+	}
```
Comment (papahmsolo) [inline tree_test1.go]: Try to `range` over the channel, it's much cleaner code

File: tree1.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
+// Package tree implements equivalence check of binary trees
+package tree
+
+import "golang.org/x/tour/tree"
```
Comment (papahmsolo) [inline tree1.go]: A little copying is better than a little dependency: https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s

Comment (papahmsolo) [reply tree1.go]: We don't really need that dependency, do we? We just need the `Tree` type.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
I have to say, tests are good

File: tree1.go (hunk 2 above)
Comment (katiasuya) [reply tree1.go]: You're right, I've complicated it one more time.

File: tree.go [hunk 3]
```diff
@@ -1,20 +1,25 @@
...
+	Left  *Tree
+	Value int
+	Right *Tree
+}
 
 // New returns a sorted binary tree holding values from nums.
-func New(nums ...int) *tree.Tree {
-	var t *tree.Tree
+func New(nums ...int) *Tree {
+	var t *Tree
```
Comment (papahmsolo) [inline tree.go:13]: We can make it a named return variable, thus we won't need the declaration.

File: tree.go [hunk 4]
```diff
@@ -1,20 +1,25 @@
...
 		t = insert(t, n)
 	}
 	return t
 }
 
-func insert(t *tree.Tree, v int) *tree.Tree {
+func insert(t *Tree, v int) *Tree {
 	if t == nil {
-		return &tree.Tree{Left: nil, Value: v, Right: nil}
+		return &Tree{Left: nil, Value: v, Right: nil}
```
Comment (papahmsolo) [inline tree.go:22]: No need to set `Left` and `Right` to `nil`, they'll get it by default

File: tree_test.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+			}
+
+			for i, n := range nums {
+				if n != tt.exp[i] {
+					t.Errorf("Expected %v, got %v", tt.exp, nums)
+					break
+				}
+			}
+		})
+	}
```
Comment (papahmsolo) [inline tree_test.go:46]: We can to this comparison easier with `reflect.DeepEqual()`

File: tree_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+// ExampleSame provides Same function example.
+func ExampleSame() {
+	fmt.Println(Same(New(0, 0, 1, 0), New(0, 0, 1, 1)),
+		Same(New(1, 2, 3, 4), New(4, 3, 1, 2)))
+	// Output: false true
+}
+
+// BenchmarkSame checks Same function's performance.
+func BenchmarkSame(b *testing.B) {
```
Comment (papahmsolo) [inline tree_test.go:113]: Let's create a benchmark for two large trees, we wat to assess the performance here

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

File: tree_test.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+			name: "multiple nodes",
+			tree: []int{8, 13, 3, 5, 1, 1, 2},
+			exp:  []int{1, 1, 2, 3, 5, 8, 13},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ch := make(chan int)
+			var nums []int
```
Comment (papahmsolo) [inline tree_test.go:32]: If we use `nums := []int{}` right away we won't need to check it for nil later

File: tree_test.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	}{
+		{
+			name: "equal",
+			in1:  []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+			in2:  []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+		},
+		{
+			name: "non-equal",
+			in1:  []int{5, -1, 9, 12, -55, 16, 2, 0, 17, -8},
+			in2:  []int{5, -9, 14, 0, 17, -1, 2, -55, 16, 4},
```
Comment (papahmsolo) [inline tree_test.go:130]: By large I mean something really large, say 1000 nodes at least. Try `rand.Perm()`, it might help you with tree generation

**Review by papahmsolo — CHANGES_REQUESTED** [review]
👍

File: tree_test.go (hunk 7 above)
Comment (katiasuya) [reply tree_test.go:32]: But what if we pass a nil value?

**Review by papahmsolo — APPROVED** [review]
Nice work!

**Pull request merged**